src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,224 @@
+        public (object? Target, object? Dependent) GetTargetAndDependent()
+        {
+            IntPtr handle = _handle;
+
+            if (handle == IntPtr.Zero)
+            {
+                ThrowHelper.ThrowInvalidOperationException();
+            }
+
+            object? target = nGetPrimaryAndSecondary(handle, out object? secondary);
+
+            return (target, secondary);
+        }
+
+        /// <summary>
+        /// Gets the target object instance for the current handle.
+        /// </summary>
+        /// <returns>The target object instance, if present.</returns>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal object? UnsafeGetTarget()
+        {
+            return nGetPrimary(_handle);
+        }
+
+        /// <summary>
+        /// Sets the target object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetTarget(object? target)
+        {
+            nSetPrimary(_handle, target);
+        }
+
+        /// <summary>
+        /// Sets the dependent object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Dependent"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetDependent(object? dependent)
+        {
+            nSetSecondary(_handle, dependent);
+        }
+
+        /// <summary>
+        /// Retrieves the values of both <see cref="Target"/> and <see cref="Dependent"/>, if available.
+        /// </summary>
+        /// <param name="dependent">The dependent instance, if available.</param>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <remarks>
+        /// This method mirrors <see cref="GetTargetAndDependent"/>, but without the allocation check.
+        /// The signature is also kept the same as the one for the internal call, to improve the codegen.
+        /// </remarks>
+        internal object? UnsafeGetTargetAndDependent(out object? dependent)
+        {
+            return nGetPrimaryAndSecondary(_handle, out dependent);
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        public void Dispose()
+        {
+            // Forces dependentHandle back to non-allocated state (if not already there)
+            // and frees the handle if needed.
+
+            if (_handle != IntPtr.Zero)
+            {
+                IntPtr handle = _handle;
+                _handle = IntPtr.Zero;
+                nFree(handle);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern IntPtr nInitialize(object? primary, object? secondary);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object? nGetPrimary(IntPtr dependentHandle);
```

Comment:
You can optimize `nGetPrimary` for CoreCLR using the same technique as what is used for `GCHandle.Target`.

Once you do that, I think we can change `nGetPrimaryAndSecondary` to just `nGetSecondary`.

Reply:
Added the optimization in bc9ee2e65c06b6766faaeef3f403bb8aa794118a, and also left a detailed comment for future reference 😄

Looking into changing that FCall now!